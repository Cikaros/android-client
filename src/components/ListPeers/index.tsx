import React, {useEffect, useState} from 'react';
import {
  FlatList,
  Image,
  Text,
  ToastAndroid,
  TouchableOpacity,
  View,
  Dimensions,
  Linking,
  StyleSheet,
} from 'react-native';
import {useSelector} from 'react-redux';
import {Peer} from '../../store/app-status/types';
import {colors} from '../../theme';
import {RootState} from 'typesafe-actions';
import {BottomSheetVirtualizedList} from '@gorhom/bottom-sheet';
import Input from '@components/Input';
import {Menu, MenuDivider, MenuItem} from 'react-native-material-menu';
import Clipboard from '@react-native-clipboard/clipboard';
import {Button} from '@components';

const windowWidth = Dimensions?.get('window').width;
const windowHeight = Dimensions?.get('window').height;

interface Props {
  bottomSheet?: boolean;
  onClose?: () => void;
}

enum StatusConnection {
  disconnected = 'disconnected',
  connected = 'connected',
}

interface PeerList extends Peer {
  key: number;
}

const ListPeers: React.FC<Props> = ({bottomSheet, onClose}) => {
  const peers = useSelector((state: RootState) => state.appStatus.peers);

  const [filteredPeers, setFilteredPeers] = useState<PeerList[]>([]);
  const [statusToFilter, setStatusToFilter] = useState<StatusConnection | null>(
    null,
  );
  const [textToFilter, setTextToFilter] = useState<string>('');
  const [filterMenuVisible, setFilterMenuVisible] = useState<boolean>(false);

  useEffect(() => {
    filterPeers(peers.data);
  }, [peers, statusToFilter, textToFilter]);

  const filterPeers = (peersToFilter: Peer[]) => {
    setFilteredPeers(
      peersToFilter
        .filter(
          (p: Peer) =>
            (statusToFilter === null ||
              p.connStatus.toLowerCase() === statusToFilter) &&
            (textToFilter.trim() === '' ||
              p.domainName.toLowerCase().includes(textToFilter.toLowerCase()) ||
              p.ip.toLowerCase().includes(textToFilter.toLowerCase())),
        )
        .map((p, i) => ({...p, key: i})),
    );
  };

  const renderItem = (peer: PeerList) => {
    const connected = 'rgba(124, 179, 5, 1)';
    const disconnected = '#aaaaaa';
    const domainName = peer.domainName;

    return (
      <TouchableOpacity
        onLongPress={() => {
          Clipboard.setString(domainName);
          ToastAndroid.show('Domain name copied!', ToastAndroid.SHORT);
        }}>
        <View
          style={{
            padding: 15,
            flex: 1,
            flexDirection: 'row',
            backgroundColor: 'rgba(0, 0, 0, 0.02)',
            width: '100%',
          }}>
          <View style={{justifyContent: 'center', marginRight: 5}}>
            <View
              style={{
                height: 40,
                borderRadius: 42,
                width: 8,
                margin: 8,
                backgroundColor:
                  peer.connStatus.toLowerCase() === StatusConnection.connected
                    ? connected
                    : disconnected,
              }}
            />
          </View>
          <View
            style={{
              flex: 1,
              flexDirection: 'column',
              justifyContent: 'center',
              gap: 8,
            }}>
            <Text style={styles.text}>{peer.domainName}</Text>
            <Text style={styles.text}>{peer.ip}</Text>
          </View>
          <View style={{justifyContent: 'center'}}>
            <Text style={[styles.text, {marginRight: 15}]}>
              {peer.connStatus.toLowerCase() === StatusConnection.connected
                ? 'connected'
                : 'disconnected'}
            </Text>
          </View>
        </View>
      </TouchableOpacity>
    );
  };

  const onChangeTextToFilter = (text: string) => {
    setTextToFilter(text);
  };

  const onSelectMenuFilter = (status: StatusConnection | null) => {
    setStatusToFilter(status);
    setFilterMenuVisible(false);
  };

  const renderMenuItem = (
    text: string,
    onPress: () => void,
    statusToEnable: StatusConnection | null,
  ) => {
    return (
      <MenuItem
        onPress={() => onPress()}
        style={{
          backgroundColor:
            statusToFilter === statusToEnable ? '#dddddd' : 'transparent',
        }}
        textStyle={{color: colors.textDefault, fontFamily: 'Roboto'}}>
        {text}
      </MenuItem>
    );
  };

  return (
    <>
      <View style={{padding: 24, marginTop: 20}}>
        <TouchableOpacity
          hitSlop={{top: 20, bottom: 20, left: 20, right: 20}}
          onPress={() => (onClose ? onClose() : false)}
          style={{
            position: 'absolute',
            right: 20,
            top: -30,
            paddingTop: 20,
            paddingLeft: 10,
            paddingRight: 10,
            paddingBottom: 10,
          }}>
          <Image
            source={require('@assets/images/close-slider-4x.png')}
            style={{width: 15, height: 15, resizeMode: 'contain'}}
          />
        </TouchableOpacity>
        {peers?.total === 0 ? (
          <View
            style={{
              flexDirection: 'column',
              alignItems: 'center',
              marginHorizontal: 8,
            }}>
            <Image
              source={require('@assets/images/icon-empty-box.png')}
              style={{
                width: windowWidth / 3,
                height: windowHeight / 7,
                resizeMode: 'contain',
              }}
            />
            <Text
              style={{
                marginBottom: 70,
                marginTop: 40,
                  color: colors.textDefault,
              }}>
              It looks like there are no machines that {'\n'} you can connect
              to...
            </Text>
            <Button
              onPress={() =>
                Linking.openURL('https://netbird.io/docs/overview/acls')
              }
              text={'Learn why'}
            />
          </View>
        ) : (
          <>
            <View
              style={{marginBottom: 32, alignItems: 'center', width: '100%'}}>
              <Text>
                <Text
                  style={{
                    fontWeight: '700',
                    color: 'rgba(0, 0, 0, 0.45)',
                    fontSize: 16,
                  }}>
                  {peers?.connected} of {peers?.total}{' '}
                </Text>
                <Text
                  style={{
                    fontWeight: '500',
                    color: 'rgba(0, 0, 0, 0.45)',
                    fontSize: 16,
                  }}>
                  Peers connected
                </Text>
              </Text>
            </View>
            <View style={{flexDirection: 'row', alignItems: 'center'}}>
              <View style={styles.mainInputStyle}>
                <Input
                  stylesMask={{borderRadius: 8}}
                  value={textToFilter}
                  placeHolder={'search peers'}
                  onChangeText={(text: string) => onChangeTextToFilter(text)}
                />
                <TouchableOpacity
                  style={{
                    position: 'absolute',
                    right: 8,
                    top: 0,
                    paddingTop: 10,
                    paddingLeft: 10,
                    paddingRight: 10,
                    paddingBottom: 10,
                  }}>
                  <Image
                    source={require('@assets/images/search-4x.png')}
                    style={{width: 15, height: 15, resizeMode: 'contain'}}
                  />
                </TouchableOpacity>
              </View>
              <View style={{flex: 0}}>
                <Menu
                  style={{marginTop: 30}}
                  visible={filterMenuVisible}
                  onRequestClose={() => setFilterMenuVisible(false)}
                  anchor={
                    <TouchableOpacity
                      style={{marginRight: 5, marginLeft: 25}}
                      hitSlop={{top: 20, bottom: 20, left: 10, right: 20}}
                      onPress={() => setFilterMenuVisible(true)}>
                      <Image
                        source={require('../../assets/images/icon-filter-4x.png')}
                        style={{width: 18, height: 18, resizeMode: 'contain'}}
                      />
                    </TouchableOpacity>
                  }>
                  {renderMenuItem('All', () => onSelectMenuFilter(null), null)}
                  <MenuDivider />
                  {renderMenuItem(
                    'Connected',
                    () => onSelectMenuFilter(StatusConnection.connected),
                    StatusConnection.connected,
                  )}
                  {renderMenuItem(
                    'Disconnected',
                    () => onSelectMenuFilter(StatusConnection.disconnected),
                    StatusConnection.disconnected,
                  )}
                </Menu>
              </View>
            </View>
          </>
        )}
      </View>
      {bottomSheet ? (
        <BottomSheetVirtualizedList
          initialNumToRender={5}
          keyExtractor={(item: PeerList) => item.key}
          style={{flex: 1, width: '100%', backgroundColor: colors.bgColor}}
          ItemSeparatorComponent={() => <View style={{height: 20}} />}
          data={filteredPeers}
          getItemCount={_data => filteredPeers.length}
          getItem={(_data, index) => filteredPeers[index]}
          renderItem={({item, index, separators}) => renderItem(item)}
        />
      ) : (
        <View
          style={{
            flex: 1,
            alignItems: 'center',
            width: '100%',
          }}>
          <FlatList
            style={{flex: 1, width: '100%', backgroundColor: colors.bgColor}}
            ItemSeparatorComponent={() => <View style={{height: 20}} />}
            data={filteredPeers}
            nestedScrollEnabled={true}
            renderItem={({item}) => renderItem(item)}
          />
        </View>
      )}
    </>
  );
};

export default React.memo(ListPeers);

const styles = StyleSheet.create({
  placeholder: {
    marginTop: 25,
    fontStyle: 'italic',
    fontSize: 28,
    color: 'red',
  },
  mainInputStyle: {
    flex: 1,
    borderRadius: 22,
  },
  text: {
    color: 'rgba(0, 0, 0, 0.45)',
    fontFamily: 'Roboto',
    fontWeight: '400',
    fontSize: 14,
  },
});
