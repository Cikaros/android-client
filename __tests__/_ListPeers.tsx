import React from 'react';
import { shallow } from 'enzyme';
import ListPeers from '../src/components/ListPeers/index'; 
import { useSelector } from 'react-redux';

// Create a mock for the useSelector hook
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
}));

// Mock the Dimensions module
jest.mock('react-native/Libraries/Utilities/Dimensions', () => ({
  get: () => ({ width: 400, height: 800 }),
}));

interface Props {
    bottomSheet: boolean,
    onClose: jest.Mock,
}

describe('ListPeers component', () => {
  // Define the default props
  const defaultProps: Props = {
    bottomSheet: true,
    onClose: jest.fn(),
  };

  // Helper function to create a shallow wrapper with the provided props
  const createWrapper = (props: Partial<Props> = {}) => {
    const finalProps = { ...defaultProps, ...props };
    return shallow(<ListPeers {...finalProps} />);
  };

  it('should render without crashing', () => {
    const wrapper = createWrapper();
    expect(wrapper.exists()).toBe(true);
  });

  it('should render the correct number of peers', () => {
    const mockPeers = [
      { domainName: 'Peer 1', ip: '192.168.0.1', connStatus: 'connected' },
      { domainName: 'Peer 2', ip: '192.168.0.2', connStatus: 'disconnected' },
    ];
    // Mock the useSelector hook to return the mockPeers
    (useSelector as jest.Mock).mockReturnValueOnce({ data: mockPeers, total: 2, connected: 1 });

    const wrapper = createWrapper();
    expect(wrapper).toMatchSnapshot()
  });
});

