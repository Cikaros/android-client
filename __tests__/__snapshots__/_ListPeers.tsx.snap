// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListPeers component should render the correct number of peers 1`] = `
<Fragment>
  <Component
    style={
      {
        "marginTop": 20,
        "padding": 24,
      }
    }
  >
    <Component
      hitSlop={
        {
          "bottom": 20,
          "left": 20,
          "right": 20,
          "top": 20,
        }
      }
      onPress={[Function]}
      style={
        {
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 20,
          "position": "absolute",
          "right": 20,
          "top": -30,
        }
      }
    >
      <Component
        source="icon-filter-4x.png"
        style={
          {
            "height": 15,
            "resizeMode": "contain",
            "width": 15,
          }
        }
      />
    </Component>
    <Component
      style={
        {
          "alignItems": "center",
          "marginBottom": 32,
          "width": "100%",
        }
      }
    >
      <Component>
        <Component
          style={
            {
              "color": "rgba(0, 0, 0, 0.45)",
              "fontSize": 16,
              "fontWeight": "700",
            }
          }
        >
          1
           of 
          2
           
        </Component>
        <Component
          style={
            {
              "color": "rgba(0, 0, 0, 0.45)",
              "fontSize": 16,
              "fontWeight": "500",
            }
          }
        >
          Peers connected
        </Component>
      </Component>
    </Component>
    <Component
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <Component
        style={
          {
            "borderRadius": 22,
            "flex": 1,
          }
        }
      >
        <Memo(Input)
          onChangeText={[Function]}
          placeHolder="search peers"
          stylesMask={
            {
              "borderRadius": 8,
            }
          }
          value=""
        />
        <Component
          style={
            {
              "paddingBottom": 10,
              "paddingLeft": 10,
              "paddingRight": 10,
              "paddingTop": 10,
              "position": "absolute",
              "right": 8,
              "top": 0,
            }
          }
        >
          <Component
            source="search-4x.png"
            style={
              {
                "height": 15,
                "resizeMode": "contain",
                "width": 15,
              }
            }
          />
        </Component>
      </Component>
      <Component
        style={
          {
            "flex": 0,
          }
        }
      >
        <Menu
          anchor={
            <UNDEFINED
              hitSlop={
                {
                  "bottom": 20,
                  "left": 10,
                  "right": 20,
                  "top": 20,
                }
              }
              onPress={[Function]}
              style={
                {
                  "marginLeft": 25,
                  "marginRight": 5,
                }
              }
            >
              <UNDEFINED
                source="icon-filter-4x.png"
                style={
                  {
                    "height": 18,
                    "resizeMode": "contain",
                    "width": 18,
                  }
                }
              />
            </UNDEFINED>
          }
          onRequestClose={[Function]}
          style={
            {
              "marginTop": 30,
            }
          }
          visible={false}
        >
          <MenuItem
            onPress={[Function]}
            style={
              {
                "backgroundColor": "#dddddd",
              }
            }
            textStyle={
              {
                "color": "rgba(77, 77, 77, 1)",
                "fontFamily": "Roboto",
              }
            }
          >
            All
          </MenuItem>
          <MenuDivider />
          <MenuItem
            onPress={[Function]}
            style={
              {
                "backgroundColor": "transparent",
              }
            }
            textStyle={
              {
                "color": "rgba(77, 77, 77, 1)",
                "fontFamily": "Roboto",
              }
            }
          >
            Connected
          </MenuItem>
          <MenuItem
            onPress={[Function]}
            style={
              {
                "backgroundColor": "transparent",
              }
            }
            textStyle={
              {
                "color": "rgba(77, 77, 77, 1)",
                "fontFamily": "Roboto",
              }
            }
          >
            Disconnected
          </MenuItem>
        </Menu>
      </Component>
    </Component>
  </Component>
  <Component
    ItemSeparatorComponent={[Function]}
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    initialNumToRender={5}
    keyExtractor={[Function]}
    renderItem={[Function]}
    style={
      {
        "backgroundColor": "#F2F2F2",
        "flex": 1,
        "width": "100%",
      }
    }
  />
</Fragment>
`;
